# NHL Skaters Multiple Linear Regression - Statsmodels

In [1]:
# import required modules
import pandas as pd
from sklearn import linear_model
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn.metrics import classification_report, confusion_matrix
import pandas as pd
from sklearn.metrics import roc_auc_score
from sklearn.metrics import precision_recall_fscore_support
import numpy as np
import statsmodels.api as sm

In [2]:
# load dataset
data = pd.read_csv('../../Capstone/Data/NHL_Spipeline1.csv', header=0)
data.shape

(43694, 27)

In [3]:
data.isna().mean().round(4)*100

Player       0.0
Tm           0.0
Pos          0.0
Season       0.0
GP           0.0
Age          0.0
G            0.0
A            0.0
PTS          0.0
PlusMinus    0.0
PIM          0.0
PS           0.0
EV           0.0
PP           0.0
SH           0.0
GW           0.0
S            0.0
BLK          0.0
HIT          0.0
FOW          0.0
FOL          0.0
EVa          0.0
PPa          0.0
SHa          0.0
S%           0.0
FO%          0.0
TOI          0.0
dtype: float64

In [4]:
data.columns

Index(['Player', 'Tm', 'Pos', 'Season', 'GP', 'Age', 'G', 'A', 'PTS',
       'PlusMinus', 'PIM', 'PS', 'EV', 'PP', 'SH', 'GW', 'S', 'BLK', 'HIT',
       'FOW', 'FOL', 'EVa', 'PPa', 'SHa', 'S%', 'FO%', 'TOI'],
      dtype='object')

In [5]:
features=['PlusMinus', 'PIM', 'SH', 'BLK', 'FOW', 'FOL', 'SHa', 'S%', 'FO%', 'TOI']
X=data[features]
y=data.PS

In [6]:
# split X and y into training and testing sets
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=2019)

In [7]:
model = sm.OLS(y, X).fit()
predictions = model.predict(X)
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                     PS   R-squared (uncentered):                   0.447
Model:                            OLS   Adj. R-squared (uncentered):              0.447
Method:                 Least Squares   F-statistic:                              3528.
Date:                Mon, 22 Jun 2020   Prob (F-statistic):                        0.00
Time:                        01:27:36   Log-Likelihood:                         -48743.
No. Observations:               43694   AIC:                                  9.751e+04
Df Residuals:                   43684   BIC:                                  9.759e+04
Df Model:                          10                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
PlusMinus      0.3061      0.004     80.553      0.000       0.299       0.314
PIM            0.3255      0.004     85.059      0.000       0.318       0.333
SH             0.0783      0.004     20.236      0.000       0.071       0.086
BLK            0.0985      0.006     16.704      0.000       0.087       0.110
FOW            0.4222      0.084      5.020      0.000       0.257       0.587
FOL           -0.0874      0.009     -9.227      0.000      -0.106      -0.069
SHa            0.1379      0.004     35.939      0.000       0.130       0.145
S%             0.2702      0.004     69.053      0.000       0.263       0.278
FO%           -0.4402      0.082     -5.385      0.000      -0.600      -0.280
TOI            0.1128      0.005     24.026      0.000       0.104       0.122
==============================================================================
Omnibus:                     2389.137   Durbin-Watson:                   1.718
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3825.812
Skew:                          -0.460   Prob(JB):                         0.00
Kurtosis:                       4.120   Cond. No.                         65.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## Equation 
   
   
PS = 0.3061(PlusMinus) + 0.3255(PIM) + 0.0783(SH) + 0.0985(BLK) + 0.4222(FOW) - 0.0874(FOL) + 0.1379(SHa) + 0.2702(S%) - 0.4402(FO%) + 0.1128(TOI)